exa/core/dao.py

```diff
@@ -0,0 +1,287 @@
+# Copyright (c) 2015-2020, Exa Analytics Development Team
+# Distributed under the terms of the Apache License 2.0
+"""
+Database access object
+####################################
+An interface that allows for composing sql queries
+with or without a sqlalchemy ORM.
+"""
+import operator
+
+from traitlets import Dict, List, Unicode, Any
+from traitlets import default, validate, TraitError
+import numpy as np
+import pandas as pd
+
+from exa.core.data import Data
+
+
+_CHUNKSIZE = 50000
+
+_op_map = {
+    k.rstrip('_'): getattr(operator, k) for k in [
+        'lt', 'le', 'eq',
+        'ne', 'ge', 'gt',
+        'not_', 'and_', 'or_'
+    ]
+}
+
+_raw_op_map = {
+    'lt': '<', 'le': '<=', 'eq': '=',
+    'ne': '!=', 'ge': '>=', 'gt': '>',
+    'not': 'not', 'and': 'and', 'or': 'or'
+}
+
+class SQLInjectionError(Exception):
+    pass
+
+
+class RawDAO(Data):
+    """The so-called database access object (DAO).
+
+    The DAO models a single table in a database.
+    Specify a schema and table_name to obtain raw
+    DAO functionality, a quick and dirty query
+    builder for low-overhead pythonic database access.
+
+    Control the columns being pulled with `entities`
+    and the amount of data to be pulled with `filters`.
+    filters is a well-formed dictionary with an
+    expected schema:
+
+    .. code-block:: Python
+
+        dao = exa.core.dao.RawDAO(table_name='users')
+        # if not specified, fetches all columns
+        dao.entities = ['id', 'user']
+        # if not specified, fetches all rows
+        dao.filters = {
+            'id': [('gt', 1), ('lt', 10)],
+            'user': [('eq', 'foo')]
+        }
+        df = dao(session=sqlalchemy.Session())
+
+    """
+    schema = Unicode(help='schema name')
+    table_name = Unicode(help='table name')
+    entities = List(help='columns to pull')
+    filters = Dict(help='filters to apply')
+
+    @default('entities')
+    def _default_entities(self):
+        return self.columns
+
+    @validate('schema', 'table_name')
+    def _validate_sql(self, prop):
+        sql = prop['value']
+        self.log.debug(f"validating {sql} against sql injection")
+        if any((i in sql for i in (';', '--'))):
+            raise SQLInjectionError("; or -- found. abort")
+        return sql
+
+    def _scan_sql(self, query):
```

Comment:
shouldn't it also include multi line (c style) comments '/*', no?

Reply:
Yeah, I will generalize the invalid token check